### ISSUE TITLE
POSTing to relay gets blocked by content blockers (again)

### ISSUE THREAD

**Issue by etergen** [getsentry/sentry#25607]
## Important Details

How are you running Sentry?
Saas (sentry.io)

## Description

Reporting to sentry.io gets blocked by content blockers. As suggested in #23054, and as previously solved in #24947, I deployed relay, which solved the issue temporarily.

[This change](https://github.com/easylist/easylist/blob/2fdf00ce423af3b880a48aefb1bd33458a8d6684/easyprivacy/easyprivacy_general.txt#L347) ([commit](https://github.com/easylist/easylist/commit/ee626674a60b59368f00557fc16e1786aa49910d)) to EasyList renders the previous fix useless.

## Steps to Reproduce

1. See the referenced commit

### What you expected to happen

Allow the `POST` to reach Sentry

I propose to construct the URL in a generic way, e.g.

- use `key` instead of `sentry_key`
- use `v` instead of `sentry_version`
- get rid of `/store/` maybe?

**Comment by etergen** [comment]
And as @BYK did last time, I'm simply pinging @getsentry/owners-ingest for triage. (special ping to @untitaker and @jan-auer)

**Comment by etergen** [comment]
> pinging @getsentry/owners-ingest for triage

That didn't work as expected... lesson learned 😄

**Comment by jan-auer** [comment]
> That didn't work as expected... lesson learned 😄

Nice try, but that one's off limits :) Let me do you the honors and route this to @getsentry/owners-ingest for you.

> This change (commit) to EasyList renders the previous fix useless.

This is particularly challenging, since this contains required information and our JavaScript SDKs send this as query parameter to prevent CORS preflights. Apart from playing a cat-and-mouse-game with EasyList/EasyPrivacy, the best immediate option I can see is to open an issue on their tracker.

**Comment by BYK** [comment]
@jan-auer I think it it reasonable to rename the GET variables?

**Comment by jan-auer** [comment]
> @jan-auer I think it it reasonable to rename the GET variables?

That's what I mean with a cat-and-mouse-game. It's a dirty workaround, until someone adds new rules. In the end, it doesn't solve the underlying issue because some people are considering Sentry a privacy concern.

**Comment by BYK** [comment]
@jan-auer agree with your point but this change will make it quite harder for them to target Sentry as we stop screaming "THIS IS SENTRYYY!!" in all requests :D

**Comment by BYK** [comment]
@etergen btw, I think you can be smart with your LB/router and remap these variables. This means you'd also need to tweak the SDKs but nobody said this was a _clean_ workaround :)

**Comment by etergen** [comment]
Thanks for the advice, and while I'm sure I'll fix it for my current scenario, I strongly believe we should provide a solution to everyone relying on Sentry, otherwise what I'll do will only serve my use case and, sadly, implementing an in-house solution, or another system would be more attractive/effective than using Sentry.

I must insist on considering one of the below alternatives:
1. Get rid of the word `sentry` from the request params (maybe use what I proposed in my first post?)
2. Move these 2 params to the request headers
3. Move these 2 params to the request payload

As @BYK correctly stated, today you are literally screaming at the world that the request is about sentry, making it easier to block. Although you won't be able to get rid of blocks on the sentry.io domain, you have everything at your disposal to avoid playing the cat-and-mouse game for those of us using the relay option.

To sum up, ideally we should be able to issue requests to a relay server in the form of `<protocol>://<server>/api/<key>` (e.g. https://example.com/api/99999999) with the DSN and sentry version inside the payload (or the headers if you prefer).

What do you think?

**Comment by nolanshaw800** [comment]
Hello! I would like to add my input that being blocked by ad blockers is an absolute **deal-breaker** with Sentry.

My company relies upon the data Sentry provides and if it continues to be being blocked by the default ublock/mainstream adblocker settings then we will need to search for another product or roll our own.

The main HTTP GET variable which easylist is using to track sentry "sentry_key" should be and/or user configurable. 

Thank you.

**Comment by BYK** [comment]
Thanks a lot for the many and good suggestions @etergen! I personally feel like there's some quick wins here and am discussing with @jan-auer and the team but at the end it will be a collaborative decision 🙂 

Thanks for adding your perspective and experience @nolanshaw800, definitely helpful to prioritize the issue.

**Comment by mitsuhiko** [comment]
Unfortunately changing the authorization scheme is not a straightforward. If we change it in all SDKs now it breaks every non sentry.io installation out there. It would have to become an opt-in mechanism or do some auto detection based on if it sends to sentry.io or a custom installation.

**Comment by BYK** [comment]
@mitsuhiko I feel like we can easily change and map server-side variable names as a start and then add a manual opt-in flag to SDKs while we figure out a good auto-negotiation way. We may use the `options` pre-flight requests.

**Comment by mitsuhiko** [comment]
We're going to discuss this in the ingest team. Going to report back with something if we have a plan.

**Comment by nolanshaw800** [comment]
It seems to me this needs to be treated like any API versioning and it wouldn't break anything if done properly.

You have authentication handshake v1 (sentry_key/etc) which will continued to be accepted and you have authentication handshake v2 which just allows optional configuration of the sentry_key query, perhaps on a project level.

**Comment by mitsuhiko** [comment]
The problem is not that versioning is hard, but that it's a thing users can easily get wrong. The likely outcome would be to have an sdk option to flip the transport format and you need to explicitly switch over to another one. But we're going to look into this. Generally speaking I'm not convinced that a simple change is not going to put us back into the same situation in a week. Easylist privacy keeps blocking even the most generic URLs. Some people had to rename their APIs away from `/events` since that is blocked.

**Comment by Spice-King** [comment]
Kinda sad to see this end up blocked a few weeks after I found Sentry, suggested it to my boss and got the green light to set aside resources for it. Domain name was easy to work around, but the `sentry_key` bit is kinda a hard blocker.

We are kinda in a somewhat unique situation where we have a number of internal services that we wanted reporting for, but also use AD policies to force install and enable u-Block Origin to cut down on IT work caused by people clicking on malicious ads.

If they start to block API paths, much like `/events` in the past, it might be an idea to fetch a JSON object with route mappings from the server from a generic endpoint say like `/api/v2/`. Something so generic that it would break other websites to block.

**Comment by intgr** [comment]
I'm a happy user of both Sentry and UBlock and I'd very much like to see Sentry crash reporting be re-enabled by ad blockers.

But to be fair, **Sentry has became a tracker** -- it ventured into the dark side by creating features like `autoSessionTracking` and enabling them by default (!!). Do most developers using Sentry even know or care about session tracking? I know I don't, but it was silently enabled behind my back simply because I unwittingly updated the Sentry SDK. 😲

This is exactly the sort of gratuitous privacy violation that EasyList intends to block, it's clear as day that the Sentry JavaScript SDK in its default configuration falls under the [EasyPrivacy policy](https://easylist.to/pages/policy.html).

Instead of behaving like EvilCorp, playing the cat-and-mouse game, setting up proxies etc, I think the only reasonable thing to do is to show some goodwill: **open a dialog with EasyList**. Communicate clearly what sort of data is collected, and ask under what conditions they are willing to lift the block. Mayhaps the session-tracking APIs can be separated from crash reporting APIs. Maybe Sentry could detect the blocking and anonymise more information in the report? etc.

**Comment by joshkel** [comment]
@intgr, it looks like the `autoSessionTracking` feature was enabled by default as part of a major (breaking) release, was documented in the changelog, and is part of the [health](https://docs.sentry.io/product/releases/health/) feature. I'm not personally using it, but I can see how it's useful. Tracking the existence of a healthy session seems different than tracking users and behavior and seems different than going to the dark side.

Sentry tried to open a dialog with EasyList (see comments from @mitsuhiko and @dcramer on #easylist/6963), but it didn't go anywhere.

**Comment by intgr** [comment]
> Sentry tried to open a dialog with EasyList (see comments from @mitsuhiko and @dcramer on #easylist/6963), but it didn't go anywhere.

That's... interesting. Maybe they didn't buy it because the communication wasn't entirely forthcoming about Sentry features?

The `autoSessionTracking` feature was [enabled by default on 2021-01-19](https://github.com/getsentry/sentry-javascript/releases/tag/6.0.0)

Literally [two days later, the CTO of Sentry](https://github.com/easylist/easylist/issues/6963#issuecomment-764837197) tries to hand-wave Sentry as "captur[ing] code diagnostics" and says "we don't provide any kind of user 'tracking' functionality whatsoever."

One could argue about the definition of "tracking", but the exchanges make it sound as if Sentry only collected information when there was a crash -- that is a half-truth at best. The reality is that in the default configuration, every pageload submits session information to Sentry.

> it looks like the autoSessionTracking feature was enabled by default as part of a major (breaking) release, was documented in the changelog

True, I accept the blame for not reading the changelog. :)

**Comment by sitompul** [comment]
hi @dcramer, will deploying an nginx proxy instance help? If it does, do you have any nginx conf recommendation?
Or do you have any workaround with relay?

**Comment by dcramer** [comment]
@sitompul we provide a proxy called Relay which would work around this: https://docs.sentry.io/product/relay/

**Comment by reuben** [comment]
Is this issue still valid? I was just looking into Relay precisely to get around default lists, but it seems it wouldn't help me at all. Our customers are relatively technical and so adblocker usage can be high, and getting vague reports about errors without being able to act on them feels like going back to the dark ages.

**Comment by notf0und** [comment]
As the OP mention, probably the best solution is to allow the sentry clients/servers to customize the name of the keys and/or allowing to be sent as part of the post request body and not as query.

The domain is not a problem since always you can use a proxy or not having sentry word on the domain when using on-premise.

With that the level of customization should be enough to avoid adblockers.

**Comment by mitsuhiko** [comment]
@notf0und the tunnel functionality allows for this: https://docs.sentry.io/platforms/javascript/troubleshooting/#using-the-tunnel-option

**Comment by sandstrom** [comment]
> Our customers are relatively technical and so adblocker usage can be high, and getting vague reports about errors without being able to act on them feels like going back to the dark ages.

We're in the same boat. Makes debugging JS errors hard.

This is the line that's blocking Sentry (in one list, at least):

https://github.com/easylist/easylist/blob/2fdf00ce/easyprivacy/easyprivacy_general.txt#L347

I did some quick research, and it seems like blocking based on an HTTP body is still rare.

https://github.com/uBlockOrigin/uBlock-issues/issues/1357

I'd suggest allowing managed relay to accept the key in either the body (one key in the JSON payload), or as a query param, and in both cases with an arbitrary name.

Relay can forward it to Sentry in the expected format. That should allow JS error reports to be received, when using Sentry with Relay.

(next level of obfuscation would probably to also base64 encode the HTTP body, and have relay also unencode it, before forwarding)
